aggregation/src/lib.rs

```diff
@@ -0,0 +1,495 @@
+            }
+        }
+    }
+}
+
+#[cfg(feature = "verify")]
+pub use verify::*;
+
+// TODO(victor): Move this into risc0-zkvm?
+/// Verifier parameters used for recursive verification (e.g. via env::verify) of receipts.
+#[derive(Clone, Default, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct RecursionVerifierParamters {
+    /// Control root to use for verifying claims via env::verify_assumption. If not provided, the
+    /// zero digest will be used, which means the same (zkVM) control root used to verify the guest
+    /// execution will be used to verify this claim.
+    pub control_root: Option<Digest>,
+}
+
+/// Input of the aggregation set builder guest.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum GuestInput {
+    /// Input is a leaf of the Merkle tree.
+    Singleton {
+        self_image_id: Digest,
+        claim: ReceiptClaim,
+    },
+    /// Input is an inner node of the Merkle tree.
+    Join {
+        self_image_id: Digest,
+        left_set_root: Digest,
+        right_set_root: Digest,
+    },
+}
+
+impl GuestInput {
+    #[inline]
+    pub fn image_id(&self) -> Digest {
+        match self {
+            GuestInput::Singleton { self_image_id, .. } => *self_image_id,
+            GuestInput::Join { self_image_id, .. } => *self_image_id,
+        }
+    }
+
+    #[inline]
+    pub fn root(&self) -> Digest {
+        match self {
+            GuestInput::Singleton { claim, .. } => claim.digest::<sha::Impl>(),
+            GuestInput::Join {
+                left_set_root,
+                right_set_root,
+                ..
+            } => commutative_keccak256(left_set_root, right_set_root),
+        }
+    }
+
+    /// Creates a [GuestOutput] from the input.
+    #[inline]
+    pub fn to_output(&self) -> GuestOutput {
+        GuestOutput::new(self.image_id(), self.root())
+    }
+}
+
+/// Calculate the Merkle root for a tree with the given list of digests as leaves.
+pub fn merkle_root(leaves: &[Digest]) -> Result<Digest, Error> {
+    match leaves {
+        [] => Err(anyhow!("Digest list is empty, cannot compute Merkle root")),
+        [digest] => Ok(*digest), // If only one digest, return it as the root
+        _ => {
+            // Split the list into two halves
+            let (left, right) = leaves.split_at(leaves.len().next_power_of_two() / 2);
+            let left_root = merkle_root(left)?;
+            let right_root = merkle_root(right)?;
+            // Hash the combined roots of the left and right halves
+            Ok(commutative_keccak256(&left_root, &right_root))
+        }
+    }
+}
+
+// TODO(victor) Should this be assembled into under a struct and impl rather than as discrete
+// functions?
+/// Calculate the Merkle path proving inclusion of the leaf at the given index in a tree
+/// constructed from the given leaves. The leaf and root are not included.
+pub fn merkle_path(leaves: &[Digest], index: usize) -> Result<Vec<Digest>, Error> {
```

Comment:
I tried this, but got an infinite loop. I spent a couple of minutes looking at it, but didn't figure out the cause.

Reply:
I could not find any reason for a infinite loop, could it be because currently the test_consistency test takes several minutes to finish (independant of the implementation)?
Anyways I've opened #338 that reduces the test_consistency and proposes to add the above changes into this branch.

Reply:
Ok, nice! I've merged #338 into this branch